Add forced progress bar option

joeyballentine opened the pull request:
<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

This PR adds the ability to force progress bars on. This allows progress bars to be fully shown in subprocesses, which allows their data to be piped. See #11171 for more details.

Closes #11171

joeyballentine committed:
Add forced progress bar option

joeyballentine committed:
add news fragment

joeyballentine committed:
Fixed issues

joeyballentine committed:
fixed flake8 problem

uranusjr commented:
I think the more common values for this are _on_ (forced on), _auto_ (detecting output format), and _off_ (never display).

joeyballentine commented:
Forced on is different from regular on though. Regular on doesn't spam stdout with messages for every progressbar update when data is being piped while running in a subprocess. Forcing it like I added, makes that happen. If that's pip's desired behavior then I could make it work that way, but I believe then all GitHub actions workflow terminals will get spammed with new progressbar lines. Therefore I believe it is better to force it explicitly

q0w commented on src/pip/_internal/cli/progress_bars.py:
```diff
@@ -47,7 +50,13 @@ def _rich_progress_bar(
+        *columns,
+        refresh_per_second=30,
+        console=Console(
+            force_terminal=bar_type == "forced",
```
It should not initialize new Console object because of missing global Console settings, like in https://github.com/pypa/pip/pull/11048
Also, your changes covers another pr https://github.com/pypa/pip/pull/11024

joeyballentine replied:
By "covers" do you just mean we will run into merge conflicts?

And I can see now that making a new console instance there could be an issue -- how do you recommend I change this option in the global console settings?

q0w replied:
> By "covers" do you just mean we will run into merge conflicts?

I think, using `auto` as a default value and `on` for forcing colors is better idea
>  how do you recommend I change this option in the global console settings?

I think you should also just use `reconfigure(force_terminal=...)`

joeyballentine replied:
I see. I will try using reconfigure.

As for it clashing with your PR, I will keep this one separate for now. If necessary I could branch off of your branch (or just redo this change once yours is merged) and just add this change there for "on" mode.

joeyballentine commented:
I also just wanna note here (I already mentioned this in #11171) that it is currently impossible to run pip in a subprocess and get the progressbar output. It will simply return the final finished progress bar. This is a known intentional thing rich does unless you force it like I added in this PR.

This is probably useful for a variety of things, but for my specific purpose I just want to parse the progress bar to show pip's download progress in my dependency manager UI. It's been a massive headache to figure all this stuff out for something that should be pretty simple. 

The workaround I've been using is to start a pip install, parse the wheel name out of the logs (using regex), terminate that subprocess, then use the pypi json API to find the download link that matches that wheel name. Then I can manually download the wheel file and track it's download progress. It's a pretty jank workaround that actually works pretty well, but I'd much rather be able to just solely use pip's built-in downloader and just parse its progress bar. But like I said, it's impossible to do at the moment without a pseudo-terminal (which at least for setting one of those up in nodejs requires compiling a bunch of stuff). 

Ideally, pip would just support the ability to force the progress bar updates to be piped through a subprocesses stdout (like I added here). Solves my issue entirely and it's barely any code changed. I tested it and it works exactly like I expected it to. 

If you have any other questions please let me know, but I'd really appreciate this being included in pip.

pfmoore commented:
Honestly, I feel that this is something that rich should handle (or more accurately, we should defer to rich on the matter). Do we know why rich works the way it currently does? I don't feel that running pip in a subprocess and relying on the data that gets written to display the progress bar is a particularly robust approach, and I'm not particularly comfortable with us enabling it unless we know rich support it (in the sense that they won't somehow change the way progress bars get rendered in a way that breaks such usage...)

joeyballentine commented:
That makes sense. An alternative that I would find useful for this case instead of doing it through rich's progress bar would be to just have an option that logs download and install progress in an easily parsable way. I don't know enough about pip's internals to be able to easily do that myself though.

joeyballentine commented:
> To be honest, if the requirement is "machine readable progress reporting", then we should look at that as a feature in its own right. It could either be a custom "progress bar style" offered by rich (somehow) or explicit code in pip to report progress to another progress somehow (probably not even linked to the progress bar code in that case).

Yeah that's kinda what I was suggesting. Something separate that reports this stuff

ichard26 commented:
Closed since we eventually added a machine readable download progress mode. https://github.com/pypa/pip/pull/12586
